File: cdm-sources/src/cdm/base/math/functions/RoundToSignificantFigures.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.math.RoundingDirectionEnum import RoundingDirectionEnum

__all__ = ['RoundToSignificantFigures']


@replaceable
def RoundToSignificantFigures(value: Decimal, significantFigures: int, roundingMode: RoundingDirectionEnum) -> Decimal:
    """
    Round a number to the supplied significant figures, using the supplied rounding direction.
    
    Parameters 
    ----------
    value : number
    The original (unrounded) number.
    
    significantFigures : int
    The number of significant figures.
    
    roundingMode : RoundingDirectionEnum
    The method of rounding (up/down/nearest).
    
    Returns
    -------
    roundedValue : number
    
    """
    _pre_registry = {}
    self = inspect.currentframe()
    
    # conditions
    
    @rosetta_local_condition(_pre_registry)
    def condition_0_NonZeroSignificantFigures(self):
        """
        The number of significant figures should be greater than zero.
        """
        return all_elements(rosetta_resolve_attr(self, "significantFigures"), ">", 0)
    # Execute all registered conditions
    execute_local_conditions(_pre_registry, 'Pre-condition')
    
    roundedValue = rosetta_resolve_attr(self, "roundedValue")
    
    
    return roundedValue

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/base/math/functions/RoundToSignificantFigures.java
package cdm.base.math.functions;

import cdm.base.math.RoundingDirectionEnum;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.ConditionValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import java.math.BigDecimal;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(RoundToSignificantFigures.RoundToSignificantFiguresDefault.class)
public abstract class RoundToSignificantFigures implements RosettaFunction {
	
	@Inject protected ConditionValidator conditionValidator;

	/**
	* @param value The original (unrounded) number.
	* @param significantFigures The number of significant figures.
	* @param roundingMode The method of rounding (up/down/nearest).
	* @return roundedValue The value to the desired number of significant figures.
	*/
	public BigDecimal evaluate(BigDecimal value, Integer significantFigures, RoundingDirectionEnum roundingMode) {
		// pre-conditions
		conditionValidator.validate(() -> greaterThan(MapperS.of(significantFigures), MapperS.of(0), CardinalityOperator.All),
			"The number of significant figures should be greater than zero.");
		
		BigDecimal roundedValue = doEvaluate(value, significantFigures, roundingMode);
		
		return roundedValue;
	}

	protected abstract BigDecimal doEvaluate(BigDecimal value, Integer significantFigures, RoundingDirectionEnum roundingMode);

	public static class RoundToSignificantFiguresDefault extends RoundToSignificantFigures {
		@Override
		protected BigDecimal doEvaluate(BigDecimal value, Integer significantFigures, RoundingDirectionEnum roundingMode) {
			BigDecimal roundedValue = null;
			return assignOutput(roundedValue, value, significantFigures, roundingMode);
		}
		
		protected BigDecimal assignOutput(BigDecimal roundedValue, BigDecimal value, Integer significantFigures, RoundingDirectionEnum roundingMode) {
			return roundedValue;
		}
	}
}
